### PR TITLE
fix(components): [carousel] Added function: height is adaptive when height is auto

### PR DESCRIPTION
**Pull request by qq282126990**
Set the height of the carousel to the height of the currently active item

### CODE REVIEW

**Comment by qq282126990** [comment]
According to the first element height
image


According to the second element height
image
image

example: 
`
<template>
  <div>
    <el-carousel height="auto" :autoplay="false" indicator-position="outside">
      <el-carousel-item v-for="item in 4" :key="item">
        <h3 class="small justify-center" text="2xl">{{ item }}</h3>
      </el-carousel-item>
    </el-carousel>
  </div>
</template>

<script setup lang="ts">
import {
    ElCarousel,
    ElCarouselItem,
} from '../../packages/components/carousel/index'
// code here
</script>

<style lang="scss">
@import '../../packages/theme-chalk/src/carousel.scss';
@import '../../packages/theme-chalk/src/carousel-item.scss';

.el-carousel__item h3 {
    color: #475669;
    opacity: 0.75;
    line-height: 150px;
    margin: 0;
    text-align: center;
}

.el-carousel__item:nth-child(2n) {
    background-color: #99a9bf;
    height: 300px;
}

.el-carousel__item:nth-child(2n + 1) {
    background-color: #d3dce6;
    height: 200px;
}
</style>

`

**Comment by qq282126990** [comment]
[Click to preview](https://element-plus.run/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cCBsYW5nPVwidHNcIj5cbmltcG9ydCB7IHJlZiwgdmVyc2lvbiBhcyB2dWVWZXJzaW9uIH0gZnJvbSAndnVlJ1xuaW1wb3J0IHsgdmVyc2lvbiBhcyBFcFZlcnNpb24gfSBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBFbGVtZW50UGx1cyB9IGZyb20gJ0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlJ1xuXG5jb25zdCBtc2cgPSByZWYoJ0hlbGxvIFdvcmxkIScpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICAgPGRpdj5cbiAgICA8ZWwtY2Fyb3VzZWwgaGVpZ2h0PVwiYXV0b1wiIDphdXRvcGxheT1cImZhbHNlXCIgaW5kaWNhdG9yLXBvc2l0aW9uPVwib3V0c2lkZVwiPlxuICAgICAgPGVsLWNhcm91c2VsLWl0ZW0gdi1mb3I9XCJpdGVtIGluIDRcIiA6a2V5PVwiaXRlbVwiPlxuICAgICAgICA8aDMgY2xhc3M9XCJzbWFsbCBqdXN0aWZ5LWNlbnRlclwiIHRleHQ9XCIyeGxcIj57eyBpdGVtIH19PC9oMz5cbiAgICAgIDwvZWwtY2Fyb3VzZWwtaXRlbT5cbiAgICA8L2VsLWNhcm91c2VsPlxuICA8L2Rpdj5cbjwvdGVtcGxhdGU+XG5cbjxzdHlsZSA+XG4uZWwtY2Fyb3VzZWxfX2l0ZW0gaDMge1xuICAgIGNvbG9yOiAjNDc1NjY5O1xuICAgIG9wYWNpdHk6IDAuNzU7XG4gICAgbGluZS1oZWlnaHQ6IDE1MHB4O1xuICAgIG1hcmdpbjogMDtcbiAgICB0ZXh0LWFsaWduOiBjZW50ZXI7XG59XG5cbi5lbC1jYXJvdXNlbF9faXRlbTpudGgtY2hpbGQoMm4pIHtcbiAgICBiYWNrZ3JvdW5kLWNvbG9yOiAjOTlhOWJmO1xuICAgIGhlaWdodDogMzAwcHg7XG59XG5cbi5lbC1jYXJvdXNlbF9faXRlbTpudGgtY2hpbGQoMm4gKyAxKSB7XG4gICAgYmFja2dyb3VuZC1jb2xvcjogI2QzZGNlNjtcbiAgICBoZWlnaHQ6IDIwMHB4O1xufVxuPC9zdHlsZT4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctOTc1NC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcInVuc3VwcG9ydGVkXCJcbiAgfVxufSIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQHZ1ZS9ydW50aW1lLWRvbUBsYXRlc3QvZGlzdC9ydW50aW1lLWRvbS5lc20tYnJvd3Nlci5qc1wiLFxuICAgIFwiQHZ1ZS9zaGFyZWRcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvc2hhcmVkQGxhdGVzdC9kaXN0L3NoYXJlZC5lc20tYnVuZGxlci5qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzXCI6IFwiaHR0cHM6Ly9wcmV2aWV3LTk3NTQtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJ1bnN1cHBvcnRlZFwiLFxuICAgIFwiQGVsZW1lbnQtcGx1cy9pY29ucy12dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0BlbGVtZW50LXBsdXMvaWNvbnMtdnVlQDIvZGlzdC9pbmRleC5taW4uanNcIlxuICB9LFxuICBcInNjb3Blc1wiOiB7fVxufSIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCB7IGdldEN1cnJlbnRJbnN0YW5jZSB9IGZyb20gJ3Z1ZSdcbmltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICByZXR1cm4gbmV3IFByb21pc2UoKHJlc29sdmUsIHJlamVjdCkgPT4ge1xuICAgIGNvbnN0IGxpbmsgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaW5rJylcbiAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgIGxpbmsuaHJlZiA9ICdodHRwczovL3ByZXZpZXctOTc1NC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmNzcydcbiAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2xvYWQnLCByZXNvbHZlKVxuICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignZXJyb3InLCByZWplY3QpXG4gICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgfSlcbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy05NzU0LWVsZW1lbnQtcGx1cy5zdXJnZS5zaC9idW5kbGUvaW5kZXguY3NzIn19)

File: packages/components/carousel/src/carousel.vue [hunk 1]
```diff
@@ -282,6 +294,10 @@ function resetTimer() {
   startTimer()
 }
 
+function setContainerHeight(height: number) {
+  containerHeight.value = height
```
Comment (tolking) [inline packages/components/carousel/src/carousel.vue:298]:
```suggestion
if (props.height === 'auto') {
  containerHeight.value = height
}
```

When height is not `auto`, extra computed should be avoided

File: packages/components/carousel/src/carousel-item.vue [hunk 2]
```diff
@@ -1,6 +1,7 @@
 <template>
   <div
     v-show="ready"
+    ref="carouselItem"
```
Comment (tolking) [inline packages/components/carousel/src/carousel-item.vue:4]:
```suggestion
    ref="carouselItemRef"
```

File: packages/components/carousel/src/carousel.vue (hunk 1 above)
Comment (qq282126990) [reply packages/components/carousel/src/carousel.vue:299]: ok , i see

**Review by tolking — COMMENTED** [review]
Could you add a test case for this?

Could  you add an example to the documentation？

**Comment by qq282126990** [comment]
i try

**Comment by qq282126990** [comment]
@tolking  There is a problem in the test case, the height of carouselItemRef.value.offsetHeight cannot be obtained

**Comment by tolking** [comment]
@qq282126990 ok

Could you add an example to the documentation？

**Comment by qq282126990** [comment]
ok i try

**Comment by qq282126990** [comment]
@tolking  
hello,examples have been added to the documentation

image

File: docs/en-US/component/carousel.md [hunk 3]
```diff
@@ -55,6 +55,16 @@ carousel/vertical
...
+## Auto height
+
+When the `height` of `carousel` is set to `auto`, the `carousel` height will be automatically set according to the height of the `carousel item`
+
+:::demo
+
+carousel/auto-height
+
+:::
+
```
Comment (tolking) [inline docs/en-US/component/carousel.md:67]: It is recommended to move the example in front of the `Card mode`.
```suggestion
## Auto height

When the `height` of `carousel` is set to `auto`, the `carousel` height will be automatically set according to the height of the `carousel item`

:::demo

carousel/auto-height

:::

## Card mode
```

File: docs/examples/carousel/auto-height.vue [hunk 4]
```diff
@@ -0,0 +1,33 @@
+<template>
+  <el-carousel height="auto" :autoplay="false">
```
Comment (tolking) [inline docs/examples/carousel/auto-height.vue:2]:
```suggestion
  <el-carousel height="auto" autoplay>
```

Set different heights for `carousel-item`

**Comment by qq282126990** [comment]
@tolking ok，I added 3 different heights

image

File: docs/en-US/component/carousel.md (hunk 3 above)
Comment (tolking) [reply docs/en-US/component/carousel.md:69]: Move the `Auto height` demo to the line 38.

File: docs/examples/carousel/auto-height.vue [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      <el-carousel-item style="height: 100px">
+        <h3 class="small justify-center" text="2xl">height 100px</h3>
+      </el-carousel-item>
+      <el-carousel-item style="height: 200px">
+        <h3 class="small justify-center" text="2xl">height 200px</h3>
+      </el-carousel-item>
+      <el-carousel-item style="height: 300px">
+        <h3 class="small justify-center" text="2xl">height 300px</h3>
+      </el-carousel-item>
+    </el-carousel>
```
Comment (tolking) [inline docs/examples/carousel/auto-height.vue:39]: I feel like an example would suffice.

Comment (tolking) [reply docs/examples/carousel/auto-height.vue:39]: Exception you need to fix the Line error

**Comment by qq282126990** [comment]
@tolking 
OK, excess examples have been deleted

image

File: docs/en-US/component/carousel.md [hunk 6]
```diff
@@ -55,6 +55,18 @@ carousel/vertical
 
 :::
 
+## Card mode
```
Comment (tolking) [inline docs/en-US/component/carousel.md:58]: Maybe there is something wrong with my statement, I recommended to move the `Auto height demo` to line 38.

Comment (qq282126990) [reply docs/en-US/component/carousel.md:58]: ok,Is that so

image

**Comment by qq282126990** [comment]
@tolking hello, It has been modified please check

**Review by tolking — APPROVED** [review]
LGTM.

**Comment by qq282126990** [comment]
@tolking 
hello, In addition, whether to consider the need for vertical card mode
image

**Comment by tolking** [comment]
There is indeed a problem with the calculation of the `card` height in `vertical` mode. you can create a new PR to fix this problem.

**Comment by qq282126990** [comment]
ok, I will modify it after this request is merged

**Pull request closed without merging**